### ISSUE TITLE
byte/strings confusion

### ISSUE THREAD

**Issue by brianmay** [cannatag/ldap3#338]
Hello,

It use to be that I would send LDAP attributes to ldap3 as `bytes` and receive them as `bytes`.

Has this changed?

It appears that I have to write attributes as `str` and I get them back as `bytes`. If I try to send `bytes` values I get errors, especially for the object class attribute.

I can provide more details on request, out of time now.

Thanks

**Comment by cannatag** [comment]
In version 2 I added an auto encoding feature (for known types) but it should accept both str and bytes and can be disable with auto_encode=False in the connection object.

**Comment by brianmay** [comment]
Byte strings received appear to have some sort of strange encoding. e.g. I am getting ``b'Linux Rules \\c2\\a3'`` when I would expect to get ``b'Linux Rules \xc2\xa3'``. As a result, I can't decode using UTF8 back the correct `str` value.

**Comment by brianmay** [comment]
> "but it should accept both str and bytes".

This is the error I was getting writing byte strings for the object classes:

```
  File "/home/brian/tree/karaage/python-tldap/.tox/py35/lib/python3.5/site-packages/ldap3/core/connection.py", line 863, in add
    raise LDAPObjectClassError('invalid object class ' + str(object_class_name))
ldap3.core.exceptions.LDAPObjectClassError: invalid object class b'organizationalUnit'
```

Which is actually appears to be two errors - one that it couldn't fine the object class (comparison with string and byte string fails), and one in that it can't generate the error string correctly.

I tried setting ``check_names`` to false, but get a similar error from another part of the code.

**Comment by brianmay** [comment]
Oh, whoops, my bad, I had hacked the code to generate the error correctly. Normally it would complain it can't add a byte string to a str.

**Comment by cannatag** [comment]
"Linux Rules \\c2\\a3" is the LDAP standard escaping format. It is just for representation and is converted to bytes when sent to the server. It is controlled by the auto_escape parameter that calls the ldap3.utils.conv.escape_filter_chars() function. This should not be used when receiving data from the server. Returned data are utf-8 encoded only if the attribute type is known to hold utf-8 encoding. Anyway you can always access the raw bytes value with the ['raw_attributes] key of the response dictionary (or the raw property of the entry).

Can you provide me some code to show the error?

Thanks,
Giovanni

**Comment by brianmay** [comment]
See test case given #339. This is only one problem mentioned here. If you fix this one, I will provide a test case for the other problem.

**Comment by cannatag** [comment]
Thanks, Will check it in the weekend.

Giovanni

**Comment by cannatag** [comment]
I've included your test in the test suite and adapted it for python 2.6 too. The tests now works fine. Can you check code in dev?

Thanks,
Giovanni

**Comment by brianmay** [comment]
Thanks! Will look at is ASAP.

**Comment by brianmay** [comment]
As far as I can tell tell, all reported issues are non-existent in the dev branch.

Assuming this isn't already done, I think it might be worthwhile writing a test case that writes unicode data to an attribute (both as a string and a byte string) and reads it back (as a byte string). However I expect the test to pass.

Thanks!

**Comment by cannatag** [comment]
I'm working to make the test suite running on Active Directory too. Will add the bytes value test soon.

Bye,
Giovanni

**Comment by brianmay** [comment]
Thanks!

**Comment by cannatag** [comment]
Hi, I've added testByteOperation.py in test. Not sure if it covers all possible byte operations. I've tested it with python 2.6, 2.7 and 3.6.
Can you have a look and check if anything is missing?

Thanks,
Giovanni

**Comment by brianmay** [comment]
Looks pretty good to me, and far more comprehensive then I imagined.

**Comment by cannatag** [comment]
good. WIll release the fix in the 2.2.3 release of ldap3.

**Comment by cannatag** [comment]
Released in ldap3 v. 2.2.3.

**Comment by brianmay** [comment]
I am still encountering issues here :-(

```
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/ldap3/core/connection.py", line 1041, in modify
    request = modify_operation(dn, changes, self.auto_encode, self.server.schema if self.server else None)
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/ldap3/operation/modify.py", line 68, in modify_operation
    partial_attribute['vals'].setComponentByPosition(index, prepare_for_sending(validate_attribute_value(schema, attribute, value, auto_encode)))
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/ldap3/protocol/convert.py", line 148, in validate_attribute_value
    raise LDAPObjectClassError('invalid class in objectClass attribute: ' + value)
TypeError: Can't convert 'bytes' object to str implicitly
```

Where value == b'top' and 'top' is in the schema. I am somewhat puzzled that I haven't seen this error until now, so will investigate from my side also.

**Comment by cannatag** [comment]
Please provide a snippet of code. Will check it immediately and release a new version if needed. I prefer to not have a 2.2.3 version still buggy.

**Comment by brianmay** [comment]
The function call that causes this is:

```python
connection.modify(
    'cn=TestProject1,ou=Group,dc=python-ldap,dc=org',
    {'objectClass': ('MODIFY_REPLACE', [b'top', b'posixGroup'])}
)
```

Does this help?

(actually now I am wondering why on earth my code even does this... still investigating)

**Comment by brianmay** [comment]
My code is correct, it was "fixing" the initial test data fixtures which were missing objectClass values. If the above in insufficient, I can provide more details as required.

**Comment by cannatag** [comment]
It's unusual to modify the objectclass attribute, so this is a case I've not checked in tests. The ldap3 library performs additional checks on the objectclass values, will try to fix it as soon as possible.

Bye,
Giovanni

**Comment by cannatag** [comment]
Please check code in dev. It should work now, even if modifying the objectClass on a real server may lead to  an objectClassViolation exception.

**Comment by brianmay** [comment]
Looks like that has fixed it. Thanks.

Yes, understand the risks of changing objectClass - especially removing classes.

**Comment by cannatag** [comment]
There are 3 kinds of object class: Structural (that defined the "nature" of the object), Auxiliary (that adds some additional attributes to the object) and Abstract (that cannot be directly instantiated, but is in the structural class hierarchy). You need always a structural class (and usually not more than one) in an object and that cannot be changed anymore. You can add and remove auxiliary classes.

**Comment by cannatag** [comment]
Fix released in ldap3 version 2.2.4
